scenes/test_string.txt

```diff
@@ -139,44 +139,134 @@
+*gosub test_finish "VAL=THERE" tokens_2
+*comment The rest of the array is wiped clean.
+*gosub test_finish "" tokens_3
+*gosub test_finish "" tokens_4
+*gosub test_finish "" tokens_5
+*gosub test_finish 5 tokens_max
+*gosub test_finish 2 tokens_count
+
+*comment Extract 6 tokens into an array of 5 (only 5 are copied).
+*gosub test_start "EXTRACT_TOKENS #3"
+*set str_buffer "A|B|C|D|E|F"
+*gosub_scene cslib_string extract_tokens str_buffer "|" "tokens"
+*comment There are actually 6 values, but only 5 are used.
+*comment The user may check for this condition.
+*gosub test_finish 6 cslib_ret
+*gosub test_finish "A" tokens_1
+*gosub test_finish "B" tokens_2
+*gosub test_finish "C" tokens_3
+*gosub test_finish "D" tokens_4
+*gosub test_finish "E" tokens_5
+*gosub test_finish 5 tokens_max
+*gosub test_finish 5 tokens_count
+
+*comment Empty source.
+*gosub test_start "EXTRACT_TOKENS #4"
+*set str_buffer ""
+*gosub_scene cslib_string extract_tokens str_buffer "|" "tokens"
+*gosub test_finish 0 cslib_ret
+*gosub test_finish "" tokens_1
+*gosub test_finish "" tokens_2
+*gosub test_finish "" tokens_3
+*gosub test_finish "" tokens_4
+*gosub test_finish "" tokens_5
+*gosub test_finish 5 tokens_max
+*gosub test_finish 0 tokens_count
+
+*comment Only one token.
+*gosub test_start "EXTRACT_TOKENS #5"
+*set str_buffer "HI"
+*gosub_scene cslib_string extract_tokens str_buffer "|" "tokens"
+*gosub test_finish 1 cslib_ret
+*gosub test_finish "HI" tokens_1
+*gosub test_finish "" tokens_2
+*gosub test_finish "" tokens_3
+*gosub test_finish "" tokens_4
+*gosub test_finish "" tokens_5
+*gosub test_finish 5 tokens_max
+*gosub test_finish 1 tokens_count
+
+*comment Empty tokens.
+*gosub test_start "EXTRACT_TOKENS #6"
+*set str_buffer "|A||C|"
+*gosub_scene cslib_string extract_tokens str_buffer "|" "tokens"
+*gosub test_finish 5 cslib_ret
+*gosub test_finish "" tokens_1
+*gosub test_finish "A" tokens_2
+*gosub test_finish "" tokens_3
+*gosub test_finish "C" tokens_4
+*gosub test_finish "" tokens_5
+*gosub test_finish 5 tokens_max
+*gosub test_finish 4 tokens_count
+
+*comment Missing separator.
+*gosub test_start "EXTRACT TOKENS #7"
+*temp result_expected "Bug: Separator should not be empty in extract_tokens."
+*set str_buffer "|A||C|"
 
+*set cslib_catch_bug true
+*gosub_scene cslib_string extract_tokens str_buffer "" "tokens"
+*set cslib_catch_bug false
+*gosub test_finish cslib_bug_message result_expected
+
+*comment Multiple separators.
+*gosub test_start "EXTRACT TOKENS #8"
+*temp result_expected "Bug: Separator should be a single char in extract_tokens."
+*set str_buffer "|A||C|"
+
+*set cslib_catch_bug true
+*gosub_scene cslib_string extract_tokens str_buffer "|-()" "tokens"
+*set cslib_catch_bug false
+*gosub test_finish cslib_bug_message result_expected
+
+*comment Empty source.
+*gosub test_start "EXTRACT_TOKENS #9"
+*set str_buffer ""
+*gosub_scene cslib_string extract_tokens str_buffer "|" "tokens"
+*gosub test_finish 0 cslib_ret
```

Comment:
Multiple finish. We could rename the two functions "test_label" "test_condition"